### PR TITLE
New File Submission Service Foundations

### PR DESCRIPTION
**Pull request by vargaat**
This PR lays down the foundations of the new file submission service codebase. I extracted various parts of the upload mechanism from UploadManager into separate classes and added unit tests to each of those. A small technical overview on how the file upload system currently looks can be found [here](https://github.com/instructure/canvas-ios/tree/feature/MBL-16145-Upload-Service-Rewrite/Core/Core/Files/FileSubmission). The follow-up of this PR will be some kind of coordinator that orchestrates the upload using classes from this PR and also connects the business logic to the UI.

refs: MBL-16145
affects: none
release note: none

test plan: none

### CODE REVIEW

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  One or more files are below the minimum test coverage 50%
  </td>
    </tr>
  
<tr>
      <td>:warning:</td>
      <td>

  This pull request will not generate a release note.
  </td>
    </tr>
  </tbody>
</table>


Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 91.56% | 91.47% | 0.08%
Core/Core/Dashboard/ViewModel/FileUploadNotificationCardListViewModel.swift | 0% | 0% | 0%
Core/Core/Dashboard/ViewModel/FileUploadNotificationCardItemViewModel.swift | 0% | 0% | 0%
Core/Core/Files/FileSubmission/Model/FileSubmissionAssembly.swift | 0% | -- | --

[MBL-16145](https://instructure.atlassian.net/browse/MBL-16145)

#### Affected Apps: None
<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: Core/Core/Files/FileSubmission/Model/Entities/FileUploadItem.swift [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    @NSManaged public var bytesUploaded: Int
+    /** The expected number of  bytes to be uploaded to finish the file upload. This can change during the upload process. */
+    @NSManaged public var bytesToUpload: Int
+    /** The file ID assigned by the API to this file after the file has been uploaded. Nil means that the file is yet to be uploaded. */
+    @NSManaged public var apiID: String?
+
+    // MARK: Upload Step All: Error
+
+    /** The description of the error happened during upload. */
+    @NSManaged public var uploadError: String?
```
Comment (szabolcs-instructure) [inline Core/Core/Files/FileSubmission/Model/Entities/FileUploadItem.swift:46]: Same as above

**Review by szabolcs-instructure — APPROVED** [review]
LGTM! I only had a few remarks

File: Core/Core/Files/FileSubmission/Model/Entities/FileSubmissionState.swift [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    }
+    var progressSum: CGFloat {
+        reduce(into: CGFloat(0)) { result, state in
+            switch state {
+            case .uploading(let progress):
+                result += progress
+            case .waiting:
+                break
+            default:
+                result += 1
```
Comment (idevainstr) [inline Core/Core/Files/FileSubmission/Model/Entities/FileSubmissionState.swift:97]: Optionally we could handle the exact cases here.

File: Core/Core/Files/FileSubmission/Model/Entities/FileSubmission.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    @NSManaged public var courseID: String
+    @NSManaged public var assignmentID: String
+    /** The user entered comment for the submission. **/
+    @NSManaged public var comment: String?
+    @NSManaged public var files: Set<FileUploadItem>
+
+    // MARK: - Submission Result
+
+    /** The description of the error happened during submission. */
+    @NSManaged public var submissionError: String?
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Files/FileSubmission/Model/Entities/FileSubmission.swift:35]: I was thinking about the same but that would mean we would have to map all the Cocoa URL errors to a human readable string which was already done at the time the Error was reported by the OS. Also API errors would need a local mapping so this felt too much work with minimal benefits.

File: Core/Core/Files/FileSubmission/Model/BackgroundURLSessionProvider.swift [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+public class BackgroundURLSessionProvider: NSObject {
+    public var session: URLSession {
+        if let activeSession = activeSession {
+            return activeSession
+        }
+        let session = createSession()
+        activeSession = session
+        return session
+    }
+    public var completionHandler: (() -> Void)?
```
[the remark this answers is not in the document]
Comment (vargaat) [reply Core/Core/Files/FileSubmission/Model/BackgroundURLSessionProvider.swift:34]: This is the closure we'll receive in the appdelegate's `application(_ application: UIApplication, handleEventsForBackgroundURLSession identifier: String, completionHandler: @escaping () -> Void)` method when the OS wakes up the app in the background.

**Review by idevainstr — APPROVED** [review]
dope! 🏅 
some optional findings...

**Review by idevainstr — APPROVED** [review]

**Pull request merged**